### PR TITLE
introduce reindex query

### PR DESCRIPTION
**Pull request by barbulescu**
Introduce query DSL for reindex source.

### CODE REVIEW

File: search-client/src/commonTest/kotlin/com/jillesvangurp/ktsearch/ReindexTest.kt [hunk 1]
```diff
@@ -72,6 +73,24 @@ class ReindexTest : SearchTestBase() {
...
     }
 
+    @Test
+    fun reindexWithQuery() = runTest {
+        client.indexDocument(sourceName, TestDocument(name = "t1"), refresh = WaitFor)
+
+        val response = client.reindex {
+            source {
+                index = sourceName
+                query = matchAll()
```
Comment (barbulescu) [inline search-client/src/commonTest/kotlin/com/jillesvangurp/ktsearch/ReindexTest.kt:83]: it provides a similar API with search and deleteByQuery

File: search-dsls/src/commonMain/kotlin/com/jillesvangurp/searchdsls/querydsl/reindex-dsl.kt [hunk 2]
```diff
@@ -33,6 +34,7 @@ class ReindexDSL : JsonDsl() {
 
 class ReindexSourceDSL : JsonDsl() {
     var index: String by property()
+    var query: ESQuery by this
```
Comment (barbulescu) [inline search-dsls/src/commonMain/kotlin/com/jillesvangurp/searchdsls/querydsl/reindex-dsl.kt:37]: A bit less code than overriding getters and setter, but we need to postprocess after block invoke.

Are setters actually used?

Comment (barbulescu) [reply search-dsls/src/commonMain/kotlin/com/jillesvangurp/searchdsls/querydsl/reindex-dsl.kt]: Introduced QueryClauses as suggested.

**Comment by jillesvangurp** [comment]
Looks good. Are you planning to do more work or is this good to go?

**Comment by barbulescu** [comment]
I think is ready now for review.

File: search-dsls/src/commonMain/kotlin/com/jillesvangurp/searchdsls/querydsl/reindex-dsl.kt [hunk 3]
```diff
@@ -31,8 +31,17 @@ class ReindexDSL : JsonDsl() {
     }
 }
 
-class ReindexSourceDSL : JsonDsl() {
+class ReindexSourceDSL : JsonDsl(), QueryClauses {
     var index: String by property()
+    var query: ESQuery
```
Comment (jillesvangurp) [inline search-dsls/src/commonMain/kotlin/com/jillesvangurp/searchdsls/querydsl/reindex-dsl.kt:36]: This should work. You could also just make query a nullable constructor parameter and deal with it in the same way as here:

```kotlin
class FilterAgg(filter: ESQuery): AggQuery("filter") {
    init {
        this[name]=filter.wrapWithName()
    }
}
```

Either is fine of course.

**Review by jillesvangurp — APPROVED** [review]

**Pull request merged**